### PR TITLE
Refactor type unification code

### PR DESCRIPTION
**Pull request by haozhun**
This prepares Presto for lambda support which requires more powerful type unification.

### CODE REVIEW

**Comment by losipiuk** [comment]
It would be nice if this one gets in after #4005 and https://github.com/Teradata/presto/pull/88.
Those PRs touch same code and are very painful to rebase as each PR contain multiple commits. 

cc: @pnowojski

File: presto-main/src/main/java/com/facebook/presto/type/TypeUnification.java [hunk 1]
```diff
@@ -0,0 +1,466 @@
...
+import java.util.Optional;
+import java.util.Queue;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static com.facebook.presto.type.TypeRegistry.isCovariantParameterPosition;
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static java.util.Objects.requireNonNull;
+
+public final class TypeUnification
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/type/TypeUnification.java]: I would very much like to see tests here.
This code support more usecases than currently accessible using SQL query syntax. 
And without UT it is hard to be sure those usecases actually work.

**Comment by martint** [comment]
@losipiuk yes, that's my current plan. I'm working my way through the other pull requests now.

**Comment by martint** [comment]
Closing for now, since @haozhun is working a new implementation.

**Pull request closed without merging**